File: Lab_Assignments/Socket/PY/tcpserver.py
import socket

HOST = "127.0.0.1"  # Localhost
PORT = 8080         # Port to listen on

# 1. Create socket (IPv4, TCP)
server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# 2. Bind to IP/Port
server_socket.bind((HOST, PORT))

# 3. Listen for connections
server_socket.listen()
print(f"TCP Server listening on {HOST}:{PORT}...")

# 4. Accept a connection
conn, addr = server_socket.accept()
print(f"Connected by {addr}")

# 5. Receive data
data = conn.recv(1024).decode()
print("Client says:", data)

# 6. Send response
conn.sendall(b"Hello from TCP server!\n")

# 7. Close sockets
conn.close()
server_socket.close()

File: Lab_Assignments/Socket/C/tcpserver.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <arpa/inet.h>

#define PORT 8080

int main() {
    int server_fd, new_socket;
    struct sockaddr_in address;
    int addrlen = sizeof(address);
    char buffer[1024] = {0};
    char *hello = "Hello from server";

    // 1. Create socket file descriptor
    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
        perror("socket failed");
        exit(EXIT_FAILURE);
    }

    // 2. Bind socket to port
    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY; // Accept connections from any IP
    address.sin_port = htons(PORT);

    if (bind(server_fd, (struct sockaddr *)&address, sizeof(address)) < 0) {
        perror("bind failed");
        exit(EXIT_FAILURE);
    }

    // 3. Listen for connections
    if (listen(server_fd, 3) < 0) {
        perror("listen");
        exit(EXIT_FAILURE);
    }

    printf("Waiting for a connection...\n");

    // 4. Accept connection
    if ((new_socket = accept(server_fd, (struct sockaddr *)&address, (socklen_t*)&addrlen)) < 0) {
        perror("accept");
        exit(EXIT_FAILURE);
    }

    // 5. Read message from client
    read(new_socket, buffer, 1024);
    printf("Message from client: %s\n", buffer);

    // 6. Send response
    send(new_socket, hello, strlen(hello), 0);
    printf("Hello message sent\n");

    // 7. Close socket
    close(new_socket);
    close(server_fd);

    return 0;
}
